datafusion/functions/src/datetime/to_local_time.rs

```diff
@@ -0,0 +1,601 @@
+            Timestamp(Microsecond, _) => Ok(Timestamp(Microsecond, None)),
+            Timestamp(Millisecond, _) => Ok(Timestamp(Millisecond, None)),
+            Timestamp(Second, _) => Ok(Timestamp(Second, None)),
+            _ => exec_err!(
+                "The to_local_time function can only accept timestamp as the arg, got {:?}", arg_types[0]
+            ),
+        }
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        if args.len() != 1 {
+            return exec_err!(
+                "to_local_time function requires 1 argument, got {:?}",
+                args.len()
+            );
+        }
+
+        self.to_local_time(args)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use arrow::array::{types::TimestampNanosecondType, TimestampNanosecondArray};
+    use arrow::compute::kernels::cast_utils::string_to_timestamp_nanos;
+    use arrow::datatypes::{DataType, TimeUnit};
+    use chrono::NaiveDateTime;
+    use datafusion_common::ScalarValue;
+    use datafusion_expr::{ColumnarValue, ScalarUDFImpl};
+
+    use super::ToLocalTimeFunc;
+
+    #[test]
+    fn test_to_local_time() {
+        let res = ToLocalTimeFunc::new().invoke(&[ColumnarValue::Scalar(
+            ScalarValue::TimestampSecond(Some(1), None),
+        )]);
+        assert!(res.is_ok());
+
+        let res = ToLocalTimeFunc::new().invoke(&[ColumnarValue::Scalar(
+            ScalarValue::TimestampSecond(Some(1), Some("+01:00".into())),
+        )]);
+        assert!(res.is_ok());
+
+        let res = ToLocalTimeFunc::new().invoke(&[ColumnarValue::Scalar(
+            ScalarValue::TimestampNanosecond(Some(1), Some("Europe/Brussels".into())),
+        )]);
+        assert!(res.is_ok());
+
+        //
+        // Fallible test cases
+        //
+
+        // invalid number of arguments -- no argument
+        let res = ToLocalTimeFunc::new().invoke(&[]);
+        assert_eq!(
+            res.err().unwrap().strip_backtrace(),
+            "Execution error: to_local_time function requires 1 argument, got 0"
+        );
+
+        // invalid number of arguments -- more than 1 argument
+        let res = ToLocalTimeFunc::new().invoke(&[
+            ColumnarValue::Scalar(ScalarValue::TimestampSecond(Some(1), None)),
+            ColumnarValue::Scalar(ScalarValue::Utf8(Some("some string".to_string()))),
+        ]);
+        assert_eq!(
+            res.err().unwrap().strip_backtrace(),
+            "Execution error: to_local_time function requires 1 argument, got 2"
+        );
+
+        // invalid argument data type
+        let res = ToLocalTimeFunc::new().invoke(&[ColumnarValue::Scalar(
+            ScalarValue::Utf8(Some("some string".to_string())),
+        )]);
+        assert_eq!(
+            res.err().unwrap().strip_backtrace(),
+            "Execution error: to_local_time function requires timestamp argument, got Utf8"
+        );
+
+        // TODO chunchun: invalid timestamp: us, ms, s
```

Comment:
is it still a TODO?

Reply:
Updated in https://github.com/apache/datafusion/pull/11347/commits/afe619f057eb4c7830f1fa01933527d21fa30d61